gradle init on Maven project results in StackOverflow

<!--- Please follow the instructions below. We receive dozens of issues every week, -->
<!--- so to stay productive, we will close issues that don't provide enough information. -->

<!--- Please open Android-related issues on the Android Issue Tracker at `https://source.android.com/source/report-bugs` -->

<!--- Provide a brief summary of the issue in the title above -->

### Expected Behavior
<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->
`gradle init` invoked on the root project of my Maven multimodule should complete without errors or fail with meaningful output

### Current Behavior
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->
`gradle init` invoked on the root project of my Maven multimodule will fail with a StackOverflow
```
bof@MBLMIXFBO /cygdrive/d/res-apps.trunk.svn/res-apps
$ gradle init --stacktrace
:wrapper UP-TO-DATE
:init
Maven to Gradle conversion is an incubating feature.
:init FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':init'.
> java.lang.StackOverflowError (no error message)

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':init'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:97)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:87)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:625)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:580)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.StackOverflowError
        at org.gradle.buildinit.plugins.internal.maven.Maven2Gradle.generateFqn(Maven2Gradle.groovy:218)
        at org.gradle.buildinit.plugins.internal.maven.Maven2Gradle.generateFqn(Maven2Gradle.groovy:223)
******** MANY MORE TIMES *******************
        at org.gradle.buildinit.plugins.internal.maven.Maven2Gradle.generateFqn(Maven2Gradle.groovy:223)
        at org.gradle.buildinit.plugins.internal.maven.Maven2Gradle.generateFqn(Maven2Gradle.groovy:223)


* Get more help at https://help.gradle.org

BUILD FAILED in 2s
2 actionable tasks: 1 executed, 1 up-to-date
```

### Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->
We have a modestly big multimodules Maven - Java/Spring project, with some (10ish) levels of pom projects. Maven builds are slow and this affects our productivity, I wanted to try gradle and I was just starting with `gradle init`.

### Steps to Reproduce (for bugs)
<!--- Provide a self-contained example project (as an attached archive or a Github project). -->
<!--- In the rare cases where this is infeasible, we will also accept a detailed set of instructions. -->
I cannot provide an example because our project is closed-sourced, and I have absolutely no hint about what caused the infinite recursion in those precise lines.
If someone can suggest me about what caused the infinite recursion I will forge a synthetic example.

### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan `https://gradle.com/scans/get-started` is ideal -->
This is my environment versions, I'll be glad to provide more info.

```
$ gradle --version

------------------------------------------------------------
Gradle 4.2.1
------------------------------------------------------------

Build time:   2017-10-02 15:36:21 UTC
Revision:     a88ebd6be7840c2e59ae4782eb0f27fbe3405ddf

Groovy:       2.4.12
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_31 (Oracle Corporation 25.31-b07)
OS:           Windows 8.1 6.3 amd64
```

--- a/subprojects/build-init/src/main/groovy/org/gradle/buildinit/plugins/internal/maven/Maven2Gradle.groovy
+++ b/subprojects/build-init/src/main/groovy/org/gradle/buildinit/plugins/internal/maven/Maven2Gradle.groovy
@@ -18,10 +18,14 @@
 
 package org.gradle.buildinit.plugins.internal.maven
 
+import groovy.util.slurpersupport.GPathResult
 import org.apache.maven.project.MavenProject
+import org.gradle.api.artifacts.ModuleIdentifier
+import org.gradle.api.internal.artifacts.DefaultModuleIdentifier
 import org.gradle.api.logging.Logger
 import org.gradle.api.logging.Logging
 import org.gradle.util.RelativePathUtil
+
 /**
  * This script obtains the effective POM of the current project, reads its dependencies
  * and generates build.gradle scripts. It also generates settings.gradle for multimodule builds. <br/>
@@ -211,18 +215,24 @@ ${globalExclusions(this.effectivePom)}
         return buffer.toString()
     }
 
-    private generateFqn(def project, def allProjects, StringBuilder buffer) {
+    private generateFqn(GPathResult project, GPathResult allProjects, StringBuilder buffer) {
         def artifactId = project.artifactId.text()
         buffer.insert(0, ":${artifactId}")
         //we don't need the top-level parent in gradle, so we stop on it
-        if (project.parent.artifactId.text() != allProjects[0].artifactId.text()) {
+        if (getModuleIdentifier(project.parent) != getModuleIdentifier(allProjects[0])) {
             def parentInBuild = allProjects.find { proj ->
-                proj.artifactId.text() == project.parent.artifactId.text()
+                getModuleIdentifier(proj) == getModuleIdentifier(project.parent)
             }
             if (parentInBuild) {
                 generateFqn(parentInBuild, allProjects, buffer)
             }
         }
+    }
+
+    private ModuleIdentifier getModuleIdentifier(GPathResult project) {
+        def artifactId = project.artifactId.text()
+        def groupId = project.groupId ? project.groupId.text() : project.parent.groupId.text()
+        return new DefaultModuleIdentifier(groupId, artifactId)
     }
 
     def localRepoUri = {

--- a/subprojects/build-init/src/test/groovy/org/gradle/buildinit/plugins/internal/maven/MavenProjectsCreatorSpec.groovy
+++ b/subprojects/build-init/src/test/groovy/org/gradle/buildinit/plugins/internal/maven/MavenProjectsCreatorSpec.groovy
@@ -141,4 +141,60 @@ class MavenProjectsCreatorSpec extends Specification {
         then:
         gradleProject.contains("compileOnly group: 'org.gradle', name: 'build-init', version:'1.0.0'")
     }
+
+    def "creates multi module project with same artifactId"() {
+        given:
+        def parentPom = temp.file("pom.xml")
+        parentPom.text = """\
+<project>
+  <modelVersion>4.0.0</modelVersion>
+  <packaging>pom</packaging>
+  <groupId>org.gradle</groupId>
+  <artifactId>test</artifactId>
+  <version>0</version>
+
+  <modules>
+    <module>commons</module>
+  </modules>
+</project>
+"""
+
+        temp.file("commons/pom.xml").text = """\
+<project>
+  <parent>
+    <groupId>org.gradle</groupId>
+    <artifactId>test</artifactId>
+    <version>0</version>
+  </parent>
+
+  <modelVersion>4.0.0</modelVersion>
+  <artifactId>commons</artifactId>
+  <packaging>pom</packaging>
+
+  <modules>
+    <module>commons</module>
+  </modules>
+
+</project>
+"""
+        temp.file("commons/commons/pom.xml").text = """\
+<project>
+  <parent>
+    <groupId>org.gradle</groupId>
+    <artifactId>commons</artifactId>
+    <version>0</version>
+  </parent>
+
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>org.gradle.commons</groupId>
+  <artifactId>commons</artifactId>
+  
+</project>
+"""
+        def mavenProjects = creator.create(settings.buildSettings(), parentPom)
+        def converter = new Maven2Gradle(mavenProjects, temp.testDirectory)
+
+        expect:
+        converter.convert()
+    }
 }

